crates/utils/src/rate_limit/rate_limiter.rs

```diff
@@ -1,15 +1,43 @@
-use crate::IpAddr;
-use std::{collections::HashMap, time::Instant};
-use strum::IntoEnumIterator;
+use enum_map::{enum_map, EnumMap};
+use once_cell::sync::Lazy;
+use std::{
+  collections::HashMap,
+  net::Ipv6Addr,
+  time::{Duration, Instant},
+};
 use tracing::debug;
 
+static START_TIME: Lazy<Instant> = Lazy::new(Instant::now);
+
+#[derive(Debug, Clone, Copy)]
+struct InstantSecs {
+  secs: u32,
+}
+
+impl InstantSecs {
+  fn now() -> Self {
+    InstantSecs {
+      secs: u32::try_from(START_TIME.elapsed().as_secs())
+        .expect("server has been running for over 136 years"),
+    }
+  }
+
+  fn secs_since(self, earlier: Self) -> u32 {
+    self.secs.saturating_sub(earlier.secs)
+  }
+
+  fn to_instant(self) -> Instant {
+    *START_TIME + Duration::from_secs(self.secs.into())
+  }
+}
+
 #[derive(Debug, Clone)]
 struct RateLimitBucket {
-  last_checked: Instant,
-  allowance: f64,
+  last_checked: InstantSecs,
+  allowance: f32,
```

Comment:
I know this code isnt well documented, but it would be helpful if you could add comments when you learn how it works.